util/closeable: add move ctor for deferred_stop

tchaikov opened the pull request:
before this change, if we move from a deferred_stop, when the original
instance is destructed, the obj referenced by it is always stopped if
the deferred_stop is not canceled. and the new instance of deferred_stop
also stops the referenced object when it is destroyed. but developer
would expect the obj is stopped only once, and should be stopped by the
new deferred_stop instance. this behavior is surprising when developer
intentially capture the deferred_stop instance by move and store it in
a local variable.

in this change,

* a move ctor is added for deferred_stop, so that the original
  instance is always canceled.
* a test is added accordingly.

please note, as a side effect, the default copy ctor is deleted. this is
expected. as deferred_stop should not be copiable. this might break
existing code. this is expected also. so developers can have a chance to
fix it using the move-ctor instead.

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

xemul commented on include/seastar/util/closeable.hh:
```diff
@@ -122,6 +122,11 @@ public:
     deferred_stop(Object& obj) noexcept : _obj(obj) {}
+    /// Moves the \c deferred_stop into a new one, and
+    /// the old one is canceled.
+    deferred_stop(deferred_stop&& x) noexcept : _obj(x._obj), _stopped(x._stopped) {
```
nit: I believe it can be a bit simpler:

```c++
deferred_stop(deferred_stop&& x) noexcept : _obj(x._obj), _stopped(std::exchange(x._stopped, true)) {}
```

tchaikov replied:
yeah. thought it was too long. but i agree it's simpler. changed as suggested.

xemul commented:
As a general comment -- the same applies to `deferred_close()` in the same .hh file, doesn't it?

avikivity commented on include/seastar/util/closeable.hh:
```diff
@@ -122,6 +122,11 @@ public:
+    /// the old one is canceled.
+    deferred_stop(deferred_stop&& x) noexcept : _obj(x._obj), _stopped(x._stopped) {
+        x.cancel();
+    }
```
It's customary to implement move-assignment if you implement move-construction, since the two are related.

tchaikov replied:
sure. added.

tchaikov commented:
changelog

- added assignment operators for both `deferred_stop` and `deferred_close`
- added a test for `deferred_close`
- mark the copy ctor `delete` -- better to be explicit than implicit.

xemul commented on include/seastar/util/closeable.hh:
```diff
@@ -61,6 +61,19 @@ public:
+    /// And the other one's state is transferred to the current one.
+    deferred_close& operator=(deferred_close&& x) noexcept {
+        do_close();
+        _obj = x._obj;
```
@tchaikov , this assignment will not re-assign and `Object& _obj` reference, but instead will try to copy referenced objects around, which is not what the `deferred_close` moving is supposed to do

tchaikov replied:
ahh, you are right. i just realized that we are not allowed to reassign a reference. see https://isocpp.org/wiki/faq/references#reseating-refs 

as an alternative, we can use `std::reference_wrapper` to hold the reference or just mark the 
```c++
deferred_close& operator=(deferred_close&& x) = delete;
```

@xemul @avikivity what do you think?

tchaikov replied:
created #1223 for further discussion.

xemul replied:
AFAIK, there's a trick

```c++
foo& operator=(foo&& o) {
    if (&o != this) {
        ~foo();
        new (this) foo(std::move(o));
    }
    return *this;
}
```

to implement move-assignment via move-ctor, but I cannot say how ... idiomatically correct it is

tchaikov replied:
hmm, i have mixed feelings here: it reuses the move ctor, so yes! but it calls destructor manually and uses placement new, this hurts the readability a little bit, IMHO.

what do you think about `std::reference_wrapper`? it's much simpler.

avikivity replied:
It's a dangerous trick. We can just change _obj to be a pointer.

tchaikov replied:
but i don't feel strong either way. please let me know what you prefer and i will update the PR accordingly if necessary.
